lib/constellation.rb

```diff
@@ -1,58 +1,31 @@
-require 'net/http'
+require "graphql/client"
+require "graphql/client/http"
 module Constellation
-  class HttpException < StandardError
+  class GraphQLException < StandardError
   end
-  HTTP_ERRORS = [
-    Errno::ETIMEDOUT,
-    Timeout::Error,
-    Net::OpenTimeout,
-    Net::ReadTimeout,
-    EOFError,
-    Errno::ECONNABORTED,
-    Errno::ECONNREFUSED,
-    Errno::ECONNRESET,
-    Errno::EHOSTDOWN,
-    Errno::EHOSTUNREACH,
-    Errno::EINVAL,
-    Errno::ENETUNREACH,
-    SocketError,
-    OpenSSL::SSL::SSLError,
-    Net::HTTPBadResponse,
-    Net::HTTPHeaderSyntaxError,
-    Net::ProtocolError,
-    Zlib::GzipFile::Error,
-  ]
 
-  def self.create_rsvp!(rsvp_params)
-    return unless constellation_enabled?
-    uri = URI.parse("#{app}/rsvps")
-    http = Net::HTTP.new(uri.host, uri.port)
-    http.use_ssl = true
-    req = Net::HTTP::Post.new(uri.path, headers)
-    req.body = { rsvp: rsvp_params.merge(total_count: true) }.to_json
-    response = http.request(req)
-    raise Constellation::HttpException, response.body unless response.code == '201'
-    return JSON.parse(response.body), response['X-Total-Count'].to_i
-  rescue *HTTP_ERRORS => e
-    raise Constellation::HttpException, e.message
-  end
-
-  def self.constellation_enabled?
-    app && token
+  HTTP = GraphQL::Client::HTTP.new("#{RsvpRails.config[:constellation_url]}/api/graphql/") do
+    def headers(_)
+      { 'Authorization' => "Bearer #{RsvpRails.config[:constellation_jwt_token]}" }
+    end
   end
+  Schema = GraphQL::Client.load_schema(HTTP)
+  Client = GraphQL::Client.new(schema: Schema, execute: HTTP)
 
-  def self.app
-    RsvpRails.config[:constellation_url]
-  end
+  CreateRsvpMutation = Client.parse <<-GRAPHQL
+    mutation($input: createRsvpInput!) {
+      createRsvp(input: $input) {
+        id
+        total_count
+      }
+    }
+  GRAPHQL
 
-  def self.token
-    RsvpRails.config[:jwt_token]
-  end
+  def self.create_rsvp!(rsvp_params)
+    rsvp_params[:event_id] = rsvp_params[:event_id].to_s
+    response = Client.query(CreateRsvpMutation, variables: { input: rsvp_params.to_h })
 
-  def self.headers
-    {
-      'Authorization' => token,
-      'Content-type' => 'application/json'
-    }
+    raise Constellation::GraphQLException, response.errors["data"] unless response.data
```

Comment:
This is standard GraphQL, but gone are the days of response code/status code checking, and here are the days of always checking `errors` (or `data`, etc).

Reply:
I guess so, judging by these docs https://github.com/github/graphql-client/blob/master/guides/handling-errors.md#parsevalidation-errors.

What interests me more in that doc is that they actually already model errors after the ActiveModel API that I mentioned https://github.com/github/graphql-client/blob/master/guides/handling-errors.md#execution-errors, so it would seem that that should indicate what fields the error occurred on?

Reply:
+1 to using `ActiveModel::Errors` so you don't have to `rescue StandardError` in the controller above. Also, this is a separate question, but how does the `graphql-client` gem handle 500s? I think we are talking about client errors here, but we may still have to check if it's not a `5xx` response (e.g. timeout).

Reply:
The service can still respond with `5xx` indeed, but it would only be about things that happen outside of query execution.